colossalai/booster/plugin/three_dim_parallel_plugin.py

```diff
@@ -0,0 +1,310 @@
+import random
+from contextlib import nullcontext
+from typing import Any, Callable, Iterator, List, Optional, Tuple, Union
+
+import numpy as np
+import torch
+import torch.distributed as dist
+from torch.distributed import ProcessGroup
+from torch.nn import Module
+from torch.optim import Optimizer
+from torch.optim.lr_scheduler import _LRScheduler as LRScheduler
+from torch.utils.data import DataLoader
+from torch.utils.data.distributed import DistributedSampler
+
+from colossalai.amp.naive_amp.mixed_precision_optimizer import MixedPrecisionOptimizer
+from colossalai.checkpoint_io import CheckpointIO
+from colossalai.cluster import ProcessGroupMesh
+from colossalai.interface import ModelWrapper, OptimizerWrapper
+from colossalai.pipeline.schedule import OneForwardOneBackwardSchedule
+from colossalai.pipeline.stage_manager import PipelineStageManager
+from colossalai.shardformer import ShardConfig, ShardFormer
+from colossalai.zero.low_level import LowLevelZeroOptimizer
+
+from .pp_plugin_base import PipelinePluginBase
+
+DP_AXIS, PP_AXIS, TP_AXIS = 0, 1, 2
+
+
+class PipelineModule(ModelWrapper):
+
+    def __init__(self, module: Module, precision: str, shard_config: ShardConfig, stage_manager: PipelineStageManager,
+                 dp_group: ProcessGroup) -> None:
+        self.stage_manager = stage_manager
+        self.dp_group = dp_group
+        shardformer = ShardFormer(shard_config)
+        module, self.shared_params = shardformer.optimize(module)
```

Comment:
Will applying model sharding affect the optimizer?

Reply:
No, it's inplace sharding, the id of parameter won't change